HM3D-M3 changes to FAST-LIO (native/ros1/src/localization/FAST_LIO).
Upstream: https://github.com/hku-mars/FAST_LIO (branch main @ 7cc4175de6f8ba2edf34bab02a42195b141027e9),
ikd-Tree: https://github.com/hku-mars/ikd-Tree @ e2e3f4e9d3b95a9e66b1ba83dc98d4a05ed8a3c4.
This patch is a modification of GPL-2.0 code and is distributed under the
GNU General Public License v2 (see the upstream LICENSE file, installed as
FAST_LIO/LICENSE). Generated by fetch_native.sh --update-patches.

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5f91a33..7df2925 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -51,7 +51,7 @@ find_package(catkin REQUIRED COMPONENTS
   std_msgs
   pcl_ros
   tf
-  livox_ros_driver
+  livox_ros_driver2
   message_generation
   eigen_conversions
 )
@@ -86,4 +86,5 @@ catkin_package(
 
 add_executable(fastlio_mapping src/laserMapping.cpp include/ikd-Tree/ikd_Tree.cpp src/preprocess.cpp)
 target_link_libraries(fastlio_mapping ${catkin_LIBRARIES} ${PCL_LIBRARIES} ${PYTHON_LIBRARIES})
-target_include_directories(fastlio_mapping PRIVATE ${PYTHON_INCLUDE_DIRS})
\ No newline at end of file
+target_include_directories(fastlio_mapping PRIVATE ${PYTHON_INCLUDE_DIRS})
+add_dependencies(fastlio_mapping ${catkin_EXPORTED_TARGETS} ${${PROJECT_NAME}_EXPORTED_TARGETS})
diff --git a/include/ikd-Tree/ikd_Tree.cpp b/include/ikd-Tree/ikd_Tree.cpp
index e8c4e86..416d253 100644
--- a/include/ikd-Tree/ikd_Tree.cpp
+++ b/include/ikd-Tree/ikd_Tree.cpp
@@ -1725,4 +1725,3 @@ bool KD_TREE<PointType>::point_cmp_z(PointType a, PointType b) { return a.z < b.
 template class KD_TREE<pcl::PointXYZ>;
 template class KD_TREE<pcl::PointXYZI>;
 template class KD_TREE<pcl::PointXYZINormal>;
-
diff --git a/include/ikd-Tree/ikd_Tree.h b/include/ikd-Tree/ikd_Tree.h
index d4b302e..e10b1ef 100644
--- a/include/ikd-Tree/ikd_Tree.h
+++ b/include/ikd-Tree/ikd_Tree.h
@@ -341,4 +341,4 @@ public:
 };
 
 // template <typename PointType>
-// PointType KD_TREE<PointType>::zeroP = PointType(0,0,0);
+// PointType KD_TREE<PointType>::zeroP = PointType(0,0,0);
\ No newline at end of file
diff --git a/package.xml b/package.xml
index 98bc774..b679528 100644
--- a/package.xml
+++ b/package.xml
@@ -12,7 +12,7 @@
 
   <maintainer email="dev@livoxtech.com">claydergc</maintainer>
 
-  <license>BSD</license>
+  <license>GPL-2.0-only</license>
 
   <author email="zhangji@cmu.edu">Ji Zhang</author>
   
@@ -25,7 +25,7 @@
   <build_depend>sensor_msgs</build_depend>
   <build_depend>tf</build_depend>
   <build_depend>pcl_ros</build_depend>
-  <build_depend>livox_ros_driver</build_depend>
+  <build_depend>livox_ros_driver2</build_depend>
   <build_depend>message_generation</build_depend>
 
   <run_depend>geometry_msgs</run_depend>
@@ -36,7 +36,7 @@
   <run_depend>std_msgs</run_depend>
   <run_depend>tf</run_depend>
   <run_depend>pcl_ros</run_depend>
-  <run_depend>livox_ros_driver</run_depend>
+  <run_depend>livox_ros_driver2</run_depend>
   <run_depend>message_runtime</run_depend>
 
   <test_depend>rostest</test_depend>
diff --git a/src/laserMapping.cpp b/src/laserMapping.cpp
index fc41408..8a0a8f7 100644
--- a/src/laserMapping.cpp
+++ b/src/laserMapping.cpp
@@ -56,7 +56,7 @@
 #include <tf/transform_datatypes.h>
 #include <tf/transform_broadcaster.h>
 #include <geometry_msgs/Vector3.h>
-#include <livox_ros_driver/CustomMsg.h>
+#include <livox_ros_driver2/CustomMsg.h>
 #include "preprocess.h"
 #include <ikd-Tree/ikd_Tree.h>
 
@@ -83,6 +83,10 @@ condition_variable sig_buffer;
 
 string root_dir = ROOT_DIR;
 string map_file_path, lid_topic, imu_topic;
+string pcd_save_dir;
+inline string pcd_save_path() {
+    return pcd_save_dir.empty() ? (string(ROOT_DIR) + "PCD/") : pcd_save_dir;
+}
 
 double res_mean_last = 0.05, total_residual = 0.0;
 double last_timestamp_lidar = 0, last_timestamp_imu = -1.0;
@@ -299,7 +303,7 @@ void standard_pcl_cbk(const sensor_msgs::PointCloud2::ConstPtr &msg)
 
 double timediff_lidar_wrt_imu = 0.0;
 bool   timediff_set_flg = false;
-void livox_pcl_cbk(const livox_ros_driver::CustomMsg::ConstPtr &msg) 
+void livox_pcl_cbk(const livox_ros_driver2::CustomMsg::ConstPtr &msg) 
 {
     mtx_buffer.lock();
     double preprocess_start_time = omp_get_wtime();
@@ -493,7 +497,7 @@ void publish_frame_world(const ros::Publisher & pubLaserCloudFull)
         sensor_msgs::PointCloud2 laserCloudmsg;
         pcl::toROSMsg(*laserCloudWorld, laserCloudmsg);
         laserCloudmsg.header.stamp = ros::Time().fromSec(lidar_end_time);
-        laserCloudmsg.header.frame_id = "camera_init";
+        laserCloudmsg.header.frame_id = "world";
         pubLaserCloudFull.publish(laserCloudmsg);
         publish_count -= PUBFRAME_PERIOD;
     }
@@ -519,7 +523,7 @@ void publish_frame_world(const ros::Publisher & pubLaserCloudFull)
         if (pcl_wait_save->size() > 0 && pcd_save_interval > 0  && scan_wait_num >= pcd_save_interval)
         {
             pcd_index ++;
-            string all_points_dir(string(string(ROOT_DIR) + "PCD/scans_") + to_string(pcd_index) + string(".pcd"));
+            string all_points_dir(pcd_save_path() + "scans_" + to_string(pcd_index) + ".pcd");
             pcl::PCDWriter pcd_writer;
             cout << "current scan saved to /PCD/" << all_points_dir << endl;
             pcd_writer.writeBinary(all_points_dir, *pcl_wait_save);
@@ -560,7 +564,7 @@ void publish_effect_world(const ros::Publisher & pubLaserCloudEffect)
     sensor_msgs::PointCloud2 laserCloudFullRes3;
     pcl::toROSMsg(*laserCloudWorld, laserCloudFullRes3);
     laserCloudFullRes3.header.stamp = ros::Time().fromSec(lidar_end_time);
-    laserCloudFullRes3.header.frame_id = "camera_init";
+    laserCloudFullRes3.header.frame_id = "world";
     pubLaserCloudEffect.publish(laserCloudFullRes3);
 }
 
@@ -569,7 +573,7 @@ void publish_map(const ros::Publisher & pubLaserCloudMap)
     sensor_msgs::PointCloud2 laserCloudMap;
     pcl::toROSMsg(*featsFromMap, laserCloudMap);
     laserCloudMap.header.stamp = ros::Time().fromSec(lidar_end_time);
-    laserCloudMap.header.frame_id = "camera_init";
+    laserCloudMap.header.frame_id = "world";
     pubLaserCloudMap.publish(laserCloudMap);
 }
 
@@ -588,7 +592,7 @@ void set_posestamp(T & out)
 
 void publish_odometry(const ros::Publisher & pubOdomAftMapped)
 {
-    odomAftMapped.header.frame_id = "camera_init";
+    odomAftMapped.header.frame_id = "world";
     odomAftMapped.child_frame_id = "body";
     odomAftMapped.header.stamp = ros::Time().fromSec(lidar_end_time);// ros::Time().fromSec(lidar_end_time);
     set_posestamp(odomAftMapped.pose);
@@ -616,14 +620,14 @@ void publish_odometry(const ros::Publisher & pubOdomAftMapped)
     q.setY(odomAftMapped.pose.pose.orientation.y);
     q.setZ(odomAftMapped.pose.pose.orientation.z);
     transform.setRotation( q );
-    br.sendTransform( tf::StampedTransform( transform, odomAftMapped.header.stamp, "camera_init", "body" ) );
+    br.sendTransform( tf::StampedTransform( transform, odomAftMapped.header.stamp, "world", "body" ) );
 }
 
 void publish_path(const ros::Publisher pubPath)
 {
     set_posestamp(msg_body_pose);
     msg_body_pose.header.stamp = ros::Time().fromSec(lidar_end_time);
-    msg_body_pose.header.frame_id = "camera_init";
+    msg_body_pose.header.frame_id = "world";
 
     /*** if path is too large, the rvis will crash ***/
     static int jjj = 0;
@@ -788,7 +792,12 @@ int main(int argc, char** argv)
     nh.param<bool>("runtime_pos_log_enable", runtime_pos_log, 0);
     nh.param<bool>("mapping/extrinsic_est_en", extrinsic_est_en, true);
     nh.param<bool>("pcd_save/pcd_save_en", pcd_save_en, false);
+    // Separate prior loading from PCD output; preserve the legacy default for other launches.
+    bool load_prior_map = false;
+    nh.param<bool>("localization/load_prior", load_prior_map, !pcd_save_en);
     nh.param<int>("pcd_save/interval", pcd_save_interval, -1);
+    nh.param<string>("pcd_save/save_dir", pcd_save_dir, "");
+    if (!pcd_save_dir.empty() && pcd_save_dir.back() != '/') pcd_save_dir += "/";
     nh.param<vector<double>>("mapping/extrinsic_T", extrinT, vector<double>());
     nh.param<vector<double>>("mapping/extrinsic_R", extrinR, vector<double>());
 
@@ -796,7 +805,7 @@ int main(int argc, char** argv)
     cout<<"p_pre->lidar_type "<<p_pre->lidar_type<<endl;
     
     path.header.stamp    = ros::Time::now();
-    path.header.frame_id ="camera_init";
+    path.header.frame_id ="world";
 
     /*** variables definition ***/
     int effect_feat_num = 0, frame_num = 0;
@@ -827,6 +836,24 @@ int main(int argc, char** argv)
     fill(epsi, epsi+23, 0.001);
     kf.init_dyn_share(get_f, df_dx, df_dw, h_share_model, NUM_MAX_ITERATIONS, epsi);
 
+    ROS_INFO("FAST-LIO mode: load_prior=%s, pcd_save=%s, max_iterations=%d",
+             load_prior_map ? "true" : "false", pcd_save_en ? "true" : "false",
+             NUM_MAX_ITERATIONS);
+    /*** Optional prior map for relocalization; normal mapping starts from scans. ***/
+    if (load_prior_map)
+    {
+        pcl::PCDReader reader;
+        string prior_map_path = pcd_save_path() + "map_kdtree.pcd";
+        if (reader.read(prior_map_path, *featsFromMap) < 0 || featsFromMap->empty())
+        {
+            ROS_FATAL("Failed to load prior map from %s", prior_map_path.c_str());
+            return -1;
+        }
+        ROS_INFO("Loaded prior map (%zu points) from %s", featsFromMap->size(), prior_map_path.c_str());
+        ikdtree.set_downsample_param(filter_size_map_min);
+        ikdtree.Build(featsFromMap->points);
+    }
+
     /*** debug record ***/
     FILE *fp;
     string pos_log_dir = root_dir + "/Log/pos_log.txt";
@@ -898,7 +925,7 @@ int main(int argc, char** argv)
             flg_EKF_inited = (Measures.lidar_beg_time - first_lidar_time) < INIT_TIME ? \
                             false : true;
             /*** Segment the map in lidar FOV ***/
-            lasermap_fov_segment();
+            // lasermap_fov_segment();
 
             /*** downsample the feature points in a scan ***/
             downSizeFilterSurf.setInputCloud(feats_undistort);
@@ -1023,11 +1050,19 @@ int main(int argc, char** argv)
     /* 2. pcd save will largely influence the real-time performences **/
     if (pcl_wait_save->size() > 0 && pcd_save_en)
     {
-        string file_name = string("scans.pcd");
-        string all_points_dir(string(string(ROOT_DIR) + "PCD/") + file_name);
         pcl::PCDWriter pcd_writer;
-        cout << "current scan saved to /PCD/" << file_name<<endl;
-        pcd_writer.writeBinary(all_points_dir, *pcl_wait_save);
+
+        string scans_dir = pcd_save_path() + "scans.pcd";
+        cout << "scans saved to " << scans_dir << endl;
+        pcd_writer.writeBinary(scans_dir, *pcl_wait_save);
+
+        PointVector ().swap(ikdtree.PCL_Storage);
+        ikdtree.flatten(ikdtree.Root_Node, ikdtree.PCL_Storage, NOT_RECORD);
+        featsFromMap->clear();
+        featsFromMap->points = ikdtree.PCL_Storage;
+        string kdtree_dir = pcd_save_path() + "map_kdtree.pcd";
+        cout << "kdtree map saved to " << kdtree_dir << endl;
+        pcd_writer.writeBinary(kdtree_dir, *featsFromMap);
     }
 
     fout_out.close();
diff --git a/src/preprocess.cpp b/src/preprocess.cpp
index 09a3959..552ea64 100644
--- a/src/preprocess.cpp
+++ b/src/preprocess.cpp
@@ -41,7 +41,7 @@ void Preprocess::set(bool feat_en, int lid_type, double bld, int pfilt_num)
   point_filter_num = pfilt_num;
 }
 
-void Preprocess::process(const livox_ros_driver::CustomMsg::ConstPtr &msg, PointCloudXYZI::Ptr &pcl_out)
+void Preprocess::process(const livox_ros_driver2::CustomMsg::ConstPtr &msg, PointCloudXYZI::Ptr &pcl_out)
 {  
   avia_handler(msg);
   *pcl_out = pl_surf;
@@ -89,7 +89,7 @@ void Preprocess::process(const sensor_msgs::PointCloud2::ConstPtr &msg, PointClo
   *pcl_out = pl_surf;
 }
 
-void Preprocess::avia_handler(const livox_ros_driver::CustomMsg::ConstPtr &msg)
+void Preprocess::avia_handler(const livox_ros_driver2::CustomMsg::ConstPtr &msg)
 {
   pl_surf.clear();
   pl_corn.clear();
diff --git a/src/preprocess.h b/src/preprocess.h
index 35f3833..b18f2b6 100644
--- a/src/preprocess.h
+++ b/src/preprocess.h
@@ -5,7 +5,7 @@
 #include <ros/ros.h>
 #include <pcl_conversions/pcl_conversions.h>
 #include <sensor_msgs/PointCloud2.h>
-#include <livox_ros_driver/CustomMsg.h>
+#include <livox_ros_driver2/CustomMsg.h>
 
 using namespace std;
 
@@ -91,7 +91,7 @@ class Preprocess
   Preprocess();
   ~Preprocess();
   
-  void process(const livox_ros_driver::CustomMsg::ConstPtr &msg, PointCloudXYZI::Ptr &pcl_out);
+  void process(const livox_ros_driver2::CustomMsg::ConstPtr &msg, PointCloudXYZI::Ptr &pcl_out);
   void process(const sensor_msgs::PointCloud2::ConstPtr &msg, PointCloudXYZI::Ptr &pcl_out);
   void set(bool feat_en, int lid_type, double bld, int pfilt_num);
 
@@ -107,7 +107,7 @@ class Preprocess
     
 
   private:
-  void avia_handler(const livox_ros_driver::CustomMsg::ConstPtr &msg);
+  void avia_handler(const livox_ros_driver2::CustomMsg::ConstPtr &msg);
   void oust64_handler(const sensor_msgs::PointCloud2::ConstPtr &msg);
   void velodyne_handler(const sensor_msgs::PointCloud2::ConstPtr &msg);
   void sim_handler(const sensor_msgs::PointCloud2::ConstPtr &msg);
